Bottom: 51be70243f0391b1703acc707a6867e1bb501795
Top:    3b43a6edeec2520949c50cd1ce3d468b5c1b95a6
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-29 20:50:57 +0100

Livedraw debug


---

diff --git a/src/autoroute.c b/src/autoroute.c
index c47c465..58c534b 100644
--- a/src/autoroute.c
+++ b/src/autoroute.c
@@ -3989,6 +3989,7 @@ struct routeone_status
   bool net_completely_routed;
 };
 
+static void ripout_livedraw_obj (routebox_t *rb);
 
 static struct routeone_status
 RouteOne (routedata_t * rd, routebox_t * from, routebox_t * to, int max_edges)
@@ -4511,6 +4512,8 @@ RouteOne (routedata_t * rd, routebox_t * from, routebox_t * to, int max_edges)
       if (rb->conflicts_with
 	  && rb->parent.expansion_area->conflicts_with != rb->conflicts_with)
 	vector_destroy (&rb->conflicts_with);
+      if (TEST_FLAG (LIVEROUTEFLAG, PCB))
+        ripout_livedraw_obj (rb);
       r_delete_entry (rd->layergrouptree[rb->group], &rb->box);
     }
   vector_destroy (&area_vec);
@@ -4607,11 +4610,13 @@ ripout_livedraw_obj (routebox_t *rb)
       LayerType *layer = LAYER_PTR (PCB->LayerGroups.Entries[rb->group][0]);
       EraseLine (rb->livedraw_obj.line);
       DestroyObject (PCB->Data, LINE_TYPE, layer, rb->livedraw_obj.line, NULL);
+      rb->livedraw_obj.line = NULL;
     }
   if (rb->type == VIA && rb->livedraw_obj.via)
     {
       EraseVia (rb->livedraw_obj.via);
       DestroyObject (PCB->Data, VIA_TYPE, rb->livedraw_obj.via, NULL, NULL);
+      rb->livedraw_obj.via = NULL;
     }
 }
 
@@ -4623,6 +4628,21 @@ ripout_livedraw_obj_cb (const BoxType * b, void *cl)
   return 0;
 }
 
+static void
+ripout_all_livedrawn (routedata_t *rd)
+{
+  routebox_t *net, *p;
+  LIST_LOOP (rd->first_net, different_net, net);
+  {
+    LIST_LOOP (net, same_net, p);
+    {
+      ripout_livedraw_obj (p);
+    }
+    END_LOOP;
+  }
+  END_LOOP;
+}
+
 struct routeall_status
 {
   /* --- for completion rate statistics ---- */
@@ -4741,8 +4761,8 @@ RouteAll (routedata_t * rd)
 		    }
 		  if (rip)
 		    {
-		      if (TEST_FLAG (LIVEROUTEFLAG, PCB))
-			ripout_livedraw_obj (p);
+                      if (TEST_FLAG (LIVEROUTEFLAG, PCB))
+                        ripout_livedraw_obj (p);
 		      del =
 			r_delete_entry (rd->layergrouptree[p->group],
 					&p->box);
@@ -5268,14 +5288,7 @@ AutoRoute (bool selected)
   changed = (RouteAll (rd).total_nets_routed > 0) || changed;
 donerouting:
   if (changed && TEST_FLAG (LIVEROUTEFLAG, PCB))
-    {
-      int i;
-      BoxType big = {0, 0, MAX_COORD, MAX_COORD};
-      for (i = 0; i < max_group; i++)
-	{
-	  r_search (rd->layergrouptree[i], &big, NULL, ripout_livedraw_obj_cb, NULL);
-	}
-    }
+    ripout_all_livedrawn (rd);
   if (changed)
     changed = IronDownAllUnfixedPaths (rd);
   Message ("Total added wire length = %f inches, %d vias added\n",
diff --git a/src/remove.c b/src/remove.c
index ab9aa5c..1b4691c 100644
--- a/src/remove.c
+++ b/src/remove.c
@@ -130,6 +130,7 @@ static void *
 DestroyVia (PinTypePtr Via)
 {
   r_delete_entry (DestroyTarget->via_tree, (BoxTypePtr) Via);
+#if 0
   free (Via->Name);
   if (Via != &DestroyTarget->Via[--DestroyTarget->ViaN])
     {
@@ -139,6 +140,7 @@ DestroyVia (PinTypePtr Via)
 		    (BoxType *) Via);
     }
   memset (&DestroyTarget->Via[DestroyTarget->ViaN], 0, sizeof (PinType));
+#endif
   return (NULL);
 }
 
@@ -149,6 +151,7 @@ static void *
 DestroyLine (LayerTypePtr Layer, LineTypePtr Line)
 {
   r_delete_entry (Layer->line_tree, (BoxTypePtr) Line);
+#if 0
   free (Line->Number);
   if (Line != &Layer->Line[--Layer->LineN])
     {
@@ -158,6 +161,7 @@ DestroyLine (LayerTypePtr Layer, LineTypePtr Line)
 		    (BoxType *) Line);
     }
   memset (&Layer->Line[Layer->LineN], 0, sizeof (LineType));
+#endif
   return (NULL);
 }
